File: python/mirage/commands/cli/builtin/git/mv.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import posixpath
from dataclasses import dataclass

from mirage.commands.cli.builtin.git.errors import (  # yapf: disable
    GitError, MoveRefusedError, MoveUsageError, NotADirectoryDestinationError,
    NoWorkspaceError, RenameFailedError, UnknownSwitchError)
from mirage.commands.cli.builtin.git.index import read_index, write_index
from mirage.commands.cli.builtin.git.io import remove_file, rename_path
from mirage.commands.cli.builtin.git.pathspec import repo_relative, under
from mirage.commands.cli.builtin.git.session import opened
from mirage.commands.cli.builtin.git.types import IndexState, RepoLocation
from mirage.commands.cli.builtin.git.util import (check_operands, fatal,
                                                  links_of, start_point)
from mirage.commands.cli.types import CLIDoors, CLIInvocation
from mirage.commands.spec.types import FlagView
from mirage.io.stream import yield_bytes
from mirage.io.types import ByteSource, IOResult
from mirage.ops.types import LinkView, StatPath
from mirage.runtime.types import DispatchFn
from mirage.types import FileStat, FileType
from mirage.utils.errors import MISS_ERRORS

# git's own wording for each way a source can be refused, in the shape
# ``fatal: <reason>, source=<src>, destination=<dst>``.
BAD_SOURCE = "bad source"
INTO_ITSELF = "can not move directory into itself"
DESTINATION_EXISTS = "destination exists"
DESTINATION_ALREADY_EXISTS = "destination already exists"
SOURCE_DIRECTORY_EMPTY = "source directory is empty"
NOT_UNDER_VERSION_CONTROL = "not under version control"


@dataclass(frozen=True, slots=True)
class MvFlags:
    """The parsed shape of a ``git mv`` invocation.

    Args:
        force (bool): ``-f``, overwrite an existing destination file.
        skip (bool): ``-k``, skip a source that cannot move rather than
            refusing the line.
        dry_run (bool): ``-n``, report what would move and move nothing.
        verbose (bool): ``-v``, print one line per move; implied by
            ``-n``.
    """
    force: bool
    skip: bool
    dry_run: bool
    verbose: bool


def parse_flags(fl: FlagView) -> MvFlags:
    """Read the raw mv flag kwargs into a frozen struct.

    Args:
        fl (FlagView): spec-validated view over the raw flag kwargs.
    """
    dry_run = fl.as_bool("dry_run")
    return MvFlags(force=fl.as_bool("force"),
                   skip=fl.as_bool("k"),
                   dry_run=dry_run,
                   verbose=fl.as_bool("verbose") or dry_run)


@dataclass(frozen=True, slots=True)
class Move:
    """One source and where it goes.

    Args:
        source (str): repository-relative path being moved.
        destination (str): repository-relative path it moves to, already
            joined with the source's basename when the destination was a
            directory.
        paths (tuple[str, ...]): the tracked paths that move with it: the
            source itself for a file, everything under it for a
            directory.
        directory (bool): whether the source is a directory.
    """
    source: str
    destination: str
    paths: tuple[str, ...]
    directory: bool


async def lstat(stat_path: StatPath, links: LinkView | None,
                path: str) -> FileStat | None:
    """What sits at a path, without following a link.

    Args:
        stat_path (StatPath): dispatcher-backed stat, both channels.
        links (LinkView | None): the name plane's link facts.
        path (str): absolute virtual path.
    """
    if links is not None:
        link = links.stat_at(path)
        if link is not None:
            return link
    return await stat_path(path)


def moved_path(move: Move, path: str) -> str:
    """Where one tracked path lands after a move.

    Args:
        move (Move): the move.
        path (str): one of ``move.paths``.
    """
    if not move.directory:
        return move.destination
    return f"{move.destination}{path[len(move.source):]}"


async def check(stat_path: StatPath, links: LinkView | None,
                location: RepoLocation, source: str, destination: str,
                tracked: set[str],
                force: bool) -> tuple[str | None, tuple[str, ...], bool]:
    """Whether one source can move, in git's own order of refusals.

    Args:
        stat_path (StatPath): dispatcher-backed stat, both channels.
        links (LinkView | None): the name plane's link facts.
        location (RepoLocation): the discovered repository.
        source (str): repository-relative source.
        destination (str): repository-relative destination.
        tracked (set[str]): repository-relative paths the index holds.
        force (bool): whether ``-f`` was given.

    Returns:
        tuple: the refusal wording or None, the tracked paths that move,
        and whether the source is a directory.
    """
    info = await lstat(stat_path, links,
                       posixpath.join(location.worktree, source))
    if info is None:
        return BAD_SOURCE, (), False
    if destination == source or destination.startswith(f"{source}/"):
        return INTO_ITSELF, (), False
    target = await lstat(stat_path, links,
                         posixpath.join(location.worktree, destination))
    if info.type is FileType.DIRECTORY:
        if target is not None:
            return DESTINATION_ALREADY_EXISTS, (), True
        inside = tuple(sorted(path for path in tracked if under(path, source)))
        if not inside:
            return SOURCE_DIRECTORY_EMPTY, (), True
        return None, inside, True
    if source not in tracked:
        return NOT_UNDER_VERSION_CONTROL, (), False
    if target is not None and (not force or target.type is FileType.DIRECTORY):
        return DESTINATION_EXISTS, (), False
    return None, (source, ), False


async def plan(stat_path: StatPath, links: LinkView | None,
               location: RepoLocation, start: str, operands: tuple[str, ...],
               tracked: set[str], flags: MvFlags) -> list[Move]:
    """Decide every move before making any, which is git's order too.

    The last operand is the destination. With several sources it has to
    be a directory that exists; with one, an existing directory takes
    the source under its own name and anything else is the new name.

    Args:
        stat_path (StatPath): dispatcher-backed stat, both channels.
        links (LinkView | None): the name plane's link facts.
        location (RepoLocation): the discovered repository.
        start (str): absolute virtual path git is running in.
        operands (tuple[str, ...]): the operands as typed.
        tracked (set[str]): repository-relative paths the index holds.
        flags (MvFlags): the parsed flags.
    """
    destination = repo_relative(location, start, operands[-1])
    target = await lstat(stat_path, links,
                         posixpath.join(location.worktree, destination))
    into = destination == "" or (target is not None
                                 and target.type is FileType.DIRECTORY)
    if len(operands) > 2 and not into:
        raise NotADirectoryDestinationError(destination)
    moves: list[Move] = []
    for operand in operands[:-1]:
        source = repo_relative(location, start, operand)
        landing = (posixpath.join(destination, posixpath.basename(source))
                   if into else destination)
        reason, paths, directory = await check(stat_path, links, location,
                                               source, landing, tracked,
                                               flags.force)
        if reason is not None:
            if flags.skip:
                continue
            raise MoveRefusedError(reason, source, landing)
        moves.append(Move(source, landing, paths, directory))
    return moves


async def apply(dispatch: DispatchFn, location: RepoLocation,
                state: IndexState, move: Move, force: bool) -> None:
    """Make one move in the working tree and the index.

    The working tree moves first, through the mount's own rename so a
    directory carries its untracked files along, and the index entries
    are re-keyed with their blob ids and modes untouched. That is what
    lets ``status`` read the result as a rename rather than a delete
    beside an add.

    Args:
        dispatch (DispatchFn): workspace op dispatcher.
        location (RepoLocation): the discovered repository.
        state (IndexState): the index, updated in place.
        move (Move): the move to make.
        force (bool): whether ``-f`` was given, which removes a file
            already at the destination first.
    """
    source = posixpath.join(location.worktree, move.source)
    destination = posixpath.join(location.worktree, move.destination)
    if force and not move.directory:
        await remove_file(dispatch, destination)
    try:
        await rename_path(dispatch, source, destination)
    except MISS_ERRORS as exc:
        raise RenameFailedError(move.source) from exc
    for path in move.paths:
        entry = state.entries.pop(path.encode())
        state.entries[moved_path(move, path).encode()] = entry


async def mv(inv: CLIInvocation[None]) -> tuple[ByteSource | None, IOResult]:
    """Move or rename a file or directory, and stage the move.

    Args:
        inv (CLIInvocation[None]): the line's invocation record.
            git declares no config_model; the planes it reads
            (data through ``dispatch``, names through ``ns``) ride
            ``inv.doors``.
    """
    doors = inv.doors or CLIDoors()
    dispatch = doors.dispatch
    stat_path = doors.stat_path
    texts = inv.texts
    fl = FlagView(inv.flags)
    try:
        if dispatch is None or stat_path is None:
            raise NoWorkspaceError()
        check_operands(texts, UnknownSwitchError)
        flags = parse_flags(fl)
        if len(texts) < 2:
            raise MoveUsageError()
        repo, location = await opened(fl, doors)
        state = await read_index(dispatch, location.gitdir)
        tracked = {
            path.decode("utf-8", errors="replace")
            for path in state.entries
        }
        moves = await plan(stat_path, links_of(doors), location,
                           start_point(fl), texts, tracked, flags)
        lines: list[str] = []
        if flags.dry_run:
            lines.extend(f"Checking rename of '{move.source}' to "
                         f"'{move.destination}'" for move in moves)
        if flags.verbose:
            lines.extend(f"Renaming {move.source} to {move.destination}"
                         for move in moves)
        if not flags.dry_run:
            for move in moves:
                await apply(dispatch, location, state, move, flags.force)
            await write_index(dispatch, location.gitdir, state)
    except GitError as exc:
        return fatal(exc)
    if not lines:
        return None, IOResult()
    return yield_bytes("".join(f"{line}\n"
                               for line in lines).encode()), IOResult()

File: typescript/packages/core/src/commands/cli/builtin/git/mv.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { IOResult } from '../../../../io/types.ts'
import type { LinkView, StatPath } from '../../../../ops/types.ts'
import { FileType, type FileStat } from '../../../../types.ts'
import { isMissingPath } from '../../../../utils/errors.ts'
import type { CommandFnResult } from '../../../config.ts'
import { FlagView } from '../../../spec/types.ts'
import type { CLIInvocation } from '../../types.ts'
import {
  GitError,
  MoveRefusedError,
  MoveUsageError,
  NotADirectoryDestinationError,
  NoWorkspaceError,
  RenameFailedError,
  UnknownSwitchError,
} from './errors.ts'
import { readIndex, updateIndex, type StagedEntry } from './index_file.ts'
import { removeFile, renamePath, under } from './io.ts'
import { basename } from './path.ts'
import { repoRelative } from './pathspec.ts'
import { opened } from './repo.ts'
import type { Dispatch, IndexEntry, RepoLocation } from './types.ts'
import { checkOperands, fatal, startPoint } from './util.ts'
import { compareCodePoints } from '../../../../utils/sort.ts'

const ENC = new TextEncoder()

// git's own wording for each way a source can be refused, in the shape
// `fatal: <reason>, source=<src>, destination=<dst>`.
const BAD_SOURCE = 'bad source'
const INTO_ITSELF = 'can not move directory into itself'
const DESTINATION_EXISTS = 'destination exists'
const DESTINATION_ALREADY_EXISTS = 'destination already exists'
const SOURCE_DIRECTORY_EMPTY = 'source directory is empty'
const NOT_UNDER_VERSION_CONTROL = 'not under version control'

/** The parsed shape of a `git mv` invocation. */
export interface MvFlags {
  /** `-f`, overwrite an existing destination file. */
  readonly force: boolean
  /** `-k`, skip a source that cannot move rather than refusing the line. */
  readonly skip: boolean
  /** `-n`, report what would move and move nothing. */
  readonly dryRun: boolean
  /** `-v`, print one line per move; implied by `-n`. */
  readonly verbose: boolean
}

/** Read the raw mv flag kwargs into a frozen struct. */
function parseFlags(fl: FlagView): MvFlags {
  const dryRun = fl.asBool('dry_run')
  return {
    force: fl.asBool('force'),
    skip: fl.asBool('k'),
    dryRun,
    verbose: fl.asBool('verbose') || dryRun,
  }
}

/** One source and where it goes. */
export interface Move {
  /** Repository-relative path being moved. */
  readonly source: string
  /**
   * Repository-relative path it moves to, already joined with the source's
   * basename when the destination was a directory.
   */
  readonly destination: string
  /**
   * The tracked paths that move with it: the source itself for a file,
   * everything under it for a directory.
   */
  readonly paths: readonly string[]
  /** Whether the source is a directory. */
  readonly directory: boolean
}

/** What one refusal check found: a reason, or the paths that move. */
interface Verdict {
  readonly reason: string | null
  readonly paths: readonly string[]
  readonly directory: boolean
}

/** What sits at a path, without following a link. */
async function lstat(
  statPath: StatPath,
  links: LinkView | null,
  path: string,
): Promise<FileStat | null> {
  const link = links?.statAt(path) ?? null
  if (link !== null) return link
  return statPath(path)
}

/** Whether a repository-relative path sits inside a directory. */
function inside(path: string, directory: string): boolean {
  return directory === '' || path.startsWith(`${directory}/`)
}

/** Where one tracked path lands after a move. */
export function movedPath(move: Move, path: string): string {
  if (!move.directory) return move.destination
  return `${move.destination}${path.slice(move.source.length)}`
}

/** Whether one source can move, in git's own order of refusals. */
export async function check(
  statPath: StatPath,
  links: LinkView | null,
  location: RepoLocation,
  source: string,
  destination: string,
  tracked: ReadonlySet<string>,
  force: boolean,
): Promise<Verdict> {
  const info = await lstat(statPath, links, under(location.worktree, source))
  if (info === null) return { reason: BAD_SOURCE, paths: [], directory: false }
  if (destination === source || destination.startsWith(`${source}/`)) {
    return { reason: INTO_ITSELF, paths: [], directory: false }
  }
  const target = await lstat(statPath, links, under(location.worktree, destination))
  if (info.type === FileType.DIRECTORY) {
    if (target !== null) return { reason: DESTINATION_ALREADY_EXISTS, paths: [], directory: true }
    const held = [...tracked].filter((path) => inside(path, source)).sort(compareCodePoints)
    if (held.length === 0) return { reason: SOURCE_DIRECTORY_EMPTY, paths: [], directory: true }
    return { reason: null, paths: held, directory: true }
  }
  if (!tracked.has(source))
    return { reason: NOT_UNDER_VERSION_CONTROL, paths: [], directory: false }
  if (target !== null && (!force || target.type === FileType.DIRECTORY)) {
    return { reason: DESTINATION_EXISTS, paths: [], directory: false }
  }
  return { reason: null, paths: [source], directory: false }
}

/**
 * Decide every move before making any, which is git's order too.
 *
 * The last operand is the destination. With several sources it has to be a
 * directory that exists; with one, an existing directory takes the source
 * under its own name and anything else is the new name.
 */
export async function plan(
  statPath: StatPath,
  links: LinkView | null,
  location: RepoLocation,
  start: string,
  operands: readonly string[],
  tracked: ReadonlySet<string>,
  flags: MvFlags,
): Promise<Move[]> {
  const destination = repoRelative(location, start, operands[operands.length - 1] ?? '')
  const target = await lstat(statPath, links, under(location.worktree, destination))
  const into = destination === '' || target?.type === FileType.DIRECTORY
  if (operands.length > 2 && !into) throw new NotADirectoryDestinationError(destination)
  const moves: Move[] = []
  for (const operand of operands.slice(0, -1)) {
    const source = repoRelative(location, start, operand)
    const landing = into
      ? destination === ''
        ? basename(source)
        : `${destination}/${basename(source)}`
      : destination
    const verdict = await check(statPath, links, location, source, landing, tracked, flags.force)
    if (verdict.reason !== null) {
      if (flags.skip) continue
      throw new MoveRefusedError(verdict.reason, source, landing)
    }
    moves.push({ source, destination: landing, paths: verdict.paths, directory: verdict.directory })
  }
  return moves
}

/**
 * Make one move in the working tree.
 *
 * The mount's own rename, so a directory carries its untracked files along.
 * Under `-f` a file already at the destination goes first, since not every
 * mount renames over one.
 */
async function apply(
  dispatch: Dispatch,
  location: RepoLocation,
  move: Move,
  force: boolean,
): Promise<void> {
  const source = under(location.worktree, move.source)
  const destination = under(location.worktree, move.destination)
  if (force && !move.directory) await removeFile(dispatch, destination)
  try {
    await renamePath(dispatch, source, destination)
  } catch (err) {
    if (isMissingPath(err)) throw new RenameFailedError(move.source)
    throw err
  }
}

/**
 * Move or rename a file or directory, and stage the move.
 *
 * The index entries are re-keyed with their blob ids and modes untouched, which
 * is what lets `status` read the result as a rename rather than a delete beside
 * an add.
 */
export async function mv(inv: CLIInvocation): Promise<CommandFnResult> {
  const doors = inv.doors ?? {}
  const texts = [...inv.texts]
  const fl = new FlagView(inv.flags)
  const lines: string[] = []
  try {
    const dispatch = doors.dispatch
    const statPath = doors.statPath
    if (statPath === undefined || dispatch === undefined) {
      throw new NoWorkspaceError()
    }
    checkOperands(texts, UnknownSwitchError)
    const flags = parseFlags(fl)
    if (texts.length < 2) throw new MoveUsageError()
    const repo = await opened(fl, doors)
    const state = await readIndex(repo, dispatch)
    const tracked = new Set(state.entries.keys())
    const moves = await plan(
      statPath,
      doors.ns?.links ?? null,
      repo.location,
      startPoint(fl),
      texts,
      tracked,
      flags,
    )
    if (flags.dryRun) {
      for (const move of moves) {
        lines.push(`Checking rename of '${move.source}' to '${move.destination}'`)
      }
    }
    if (flags.verbose) {
      for (const move of moves) lines.push(`Renaming ${move.source} to ${move.destination}`)
    }
    if (!flags.dryRun) {
      const staged = new Map<string, StagedEntry>()
      const removed: string[] = []
      for (const move of moves) {
        await apply(dispatch, repo.location, move, flags.force)
        for (const path of move.paths) {
          const entry: IndexEntry | undefined = state.entries.get(path)
          if (entry === undefined) continue
          removed.push(path)
          staged.set(movedPath(move, path), { oid: entry.oid, mode: entry.mode, size: entry.size })
        }
      }
      await updateIndex(repo, staged, removed)
    }
  } catch (err) {
    if (err instanceof GitError) return fatal(err)
    throw err
  }
  if (lines.length === 0) return [null, new IOResult()]
  return [ENC.encode(lines.map((line) => `${line}\n`).join('')), new IOResult()]
}
